tasks/image_size.js

```diff
@@ -7,53 +7,92 @@
-			if ((!options.configObject || !options.configObject.length) && !file.dest) {
-				grunt.fail.warn('Neither options.configObject nor dest specified for this task.');
-				return;
-			}
-
-			var output = [];
-			file.src.forEach(function (src) {
-				if (grunt.file.exists(src) && grunt.file.isFile(src) && types.indexOf(path.extname(src).toLowerCase()) !== -1) {
-					var dimensions = sizer(src);
-					output.push({
-						name : src,
-						width : dimensions.width,
-						height : dimensions.height
-					});
-
-					grunt.verbose.writeln('dimensions of ' + src + ' width: ' + dimensions.width.toString() + ', height: ' + dimensions.height.toString());
-				}
-			});
-
-			if (options.configObject && options.configObject.length) {
-				grunt.config.set(options.configObject, output);
-			}
-
-			if (file.dest) {
-				grunt.file.write(file.dest, JSON.stringify(output));
-				grunt.log.writeln('File ' + chalk.cyan(file.dest) + ' created.');
-			}
-		});
-	});
+  grunt.registerMultiTask(
+    'image_size',
+    'Retrieve images size information',
+    function() {
+      var options = this.options({
+        processName: undefined,
+        processEntry: undefined,
+        processSizes: undefined,
+        configObject: undefined,
+        replacer: null,
+        space: 2
+      });
+
+      if (!this.files.length) {
+        return grunt.log.error('No files or `options.configObject` specified.');
+      }
+
+      var processedFiles = 0;
+
+      this.files.forEach(function(file) {
+        if (!file.src.length)
+          return grunt.log.error(
+            'No source files specified for ' + chalk.cyan(dest) + '.'
+          );
+
+        if (!file.dest) return grunt.log.error('No dest file specified.');
+
+        var sizes = [];
+
+        file.src.forEach(function(src) {
+          if (!grunt.file.isFile(src)) return;
+
+          var size = sizer(src);
+          var name = src;
+
+          if (typeof options.processName === 'function') {
+            name = options.processName.call(file, src, file);
+          }
+
+          var entry = { name: name, width: size.width, height: size.height };
+
+          if (typeof options.processEntry === 'function') {
+            entry = options.processEntry.call(file, entry, src, file);
+          }
+
+          sizes.push(entry);
+
+          grunt.verbose.writeln(
+            'Size of ' +
+              src +
+              ' width: ' +
+              size.width.toString() +
+              ', height: ' +
+              size.height.toString()
+          );
+
+          processedFiles++;
+        });
+
+        if (typeof options.processSizes === 'function') {
+          sizes = options.processSizes.call(file, sizes, file);
+        }
+
+        if (options.configObject && options.configObject.length) {
+          grunt.config.set(options.configObject, sizes);
+        }
+
+        grunt.file.write(
+          file.dest,
+          JSON.stringify(sizes, options.replacer, options.space)
```

Comment:
What if processSizes returns string (or Buffer), with prepared output, so we do not need replacer and space options at all. And user not restricted with json format for output.

Reply:
Hm, good catch.

Right now task assumes that `processSizes` can return only serializable data, and nothing more.

We can make something like this:

```diff
      var options = this.options({
+       processSizes: function (sizes) {
+         return JSON.stringify(sizes, null, 2);
+       },
        replacer: null,
        space: 2
      });

...

-       if (typeof options.processSizes === 'function') {
          sizes = options.processSizes.call(file, sizes, file);
-       }

        if (shouldSetConfigObject) {
          grunt.config.set(options.configObject, sizes);
        }

        if (file.dest) {
          grunt.file.write(
            file.dest,
-           JSON.stringify(sizes, options.replacer, options.space)
+           sizes
          );

          grunt.log.writeln('File ' + chalk.cyan(file.dest) + ' created.');
        }

```

It means that `processSizes` now will have predefined value, and by default will stringify input into the file, but the user can override it.

Though, it is quite new for Grunt task configs, not a common thing. I have a feeling that it might make a configuration not that clear for newcomers.

Reply:
Agree, leave replacer and space options

Reply:
I generally like the idea that it might be possible to expand the output beyond serializable data. But yeap, let's leave it for the future when someones really would need it.